app/src/main/java/net/aiscope/gdd_app/ui/mask/MaskActivity.kt

```diff
@@ -90,20 +93,45 @@ class MaskActivity : AppCompatActivity(), MaskView, CaptureFlow {
     override fun loadBitmap(imagePath: String) {
         coroutineScope.launch {
             val bmp = readImage(imagePath)
-            maskView.setImageBitmap(bmp)
+            mask_custom_view.setImageBitmap(bmp)
         }
     }
 
-    override fun brushMode() {
-        maskView.mode = MaskCustomView.DrawMode.Brush
+    override fun zoomMode() = mask_custom_view.zoomMode()
+
+    override fun drawMode() = mask_custom_view.drawMode()
+
+    override fun drawUndo() {
+        mask_custom_view.undo()
+        setEnabled(undo_btn, mask_custom_view.undoAvailable())
+        setEnabled(redo_btn, true)
+    }
+
+    override fun drawRedo() {
+        mask_custom_view.redo()
+        setEnabled(undo_btn, true)
+        setEnabled(redo_btn, mask_custom_view.redoAvailable())
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        outState.putInt("undo_btn-visibility", undo_btn.visibility)
+        outState.putInt("redo_btn-visibility", redo_btn.visibility)
+    }
+
+    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
+        super.onRestoreInstanceState(savedInstanceState)
+        undo_btn.visibility = savedInstanceState.getInt("undo_btn-visibility", View.INVISIBLE)
+        redo_btn.visibility = savedInstanceState.getInt("redo_btn-visibility", View.INVISIBLE)
     }
 
-    override fun eraseMode() {
-        maskView.mode = MaskCustomView.DrawMode.Erase
+    private fun refreshButtonsVisibility() {
+        setEnabled(undo_btn, mask_custom_view.undoAvailable())
+        setEnabled(redo_btn, mask_custom_view.redoAvailable())
     }
 
-    override fun moveMode() {
-        maskView.mode = MaskCustomView.DrawMode.Move
+    private fun setEnabled(view: View, enabled: Boolean) {
```

Comment:
Just a nitpick, but I am not convinced that this is the right name for this method, maybe something like `setVisible` instead? 
Also, I think that this is a very good use case for Kotlin's extension methods 👀